### ISSUE TITLE
Compiling asset targets being run multiple times

### ISSUE THREAD

**Issue by Therzok** [xamarin/xamarin-macios#13544]
### Steps to Reproduce

1. Solution with library projects targeting net6.0-macos with no assets, just using Xamarin.Mac reference (will prepare sample)
2. In incremental builds, seeing duplicate targets being skipped, but that means they are called multiple times.

```
   135>_CoreCompileImageAssets:
       Skipping target "_CoreCompileImageAssets" because all output files are up-to-date with respect to the input files.
       _BeforeCoreCompileInterfaceDefinitions:
       Skipping target "_BeforeCoreCompileInterfaceDefinitions" because it has no inputs.
       _BeforeCoreCompileInterfaceDefinitions:
       Skipping target "_BeforeCoreCompileInterfaceDefinitions" because it has no inputs.
       _BeforeCoreCompileInterfaceDefinitions:
       Skipping target "_BeforeCoreCompileInterfaceDefinitions" because it has no inputs.
       _CoreCompileInterfaceDefinitions:
       Skipping target "_CoreCompileInterfaceDefinitions" because it has no inputs.
       _BeforeCoreCompileSceneKitAssets:
       Skipping target "_BeforeCoreCompileSceneKitAssets" because it has no inputs.
       _BeforeCoreCompileSceneKitAssets:
       Skipping target "_BeforeCoreCompileSceneKitAssets" because it has no inputs.
       _BeforeCoreCompileSceneKitAssets:
       Skipping target "_BeforeCoreCompileSceneKitAssets" because it has no inputs.
       _CoreCompileSceneKitAssets:
       Skipping target "_CoreCompileSceneKitAssets" because it has no inputs.
       _BeforeCompileTextureAtlases:
       Skipping target "_BeforeCompileTextureAtlases" because it has no inputs.
       _BeforeCompileTextureAtlases:
         Skipping target "_BeforeCompileTextureAtlases" because it has no inputs.
       _BeforeCompileTextureAtlases:
       Skipping target "_BeforeCompileTextureAtlases" because it has no inputs.
       _CoreCompileTextureAtlases:
       Skipping target "_CoreCompileTextureAtlases" because it has no inputs.
       _BeforeCompileCoreMLModels:
       Skipping target "_BeforeCompileCoreMLModels" because all output files are up-to-date with respect to the input files.
       _CoreCompileCoreMLModels:
       Skipping target "_CoreCompileCoreMLModels" because all output files are up-to-date with respect to the input files.
```

Looking at the definitions, at least for CompileTextureAtlases:
https://github.com/xamarin/xamarin-macios/blob/b8e9c61409553c5ed94b9a8c08197ac2b0da4bef/msbuild/Xamarin.Shared/Xamarin.Shared.targets#L1081-L1088

`_BeforeCompileTextureAtlases` is mentioned in `_CompileTextureAtlasesDependsOn`, and in the DependsOnTargets of both `_ReadCoreCompileTextureAtlases` and `_CoreCompileTextureAtlases `. Normally, a target should be run only once, so I'm not sure what's going on here.

https://github.com/xamarin/xamarin-macios/blob/b8e9c61409553c5ed94b9a8c08197ac2b0da4bef/msbuild/Xamarin.Shared/Xamarin.Shared.targets#L1116-L1117

https://github.com/xamarin/xamarin-macios/blob/b8e9c61409553c5ed94b9a8c08197ac2b0da4bef/msbuild/Xamarin.Shared/Xamarin.Shared.targets#L1126-L1129

The same pattern applies for `_BeforeCoreCompileSceneKitAssets`.

I can share a binlog over sharepoint if needed.

### Expected Behavior

Targets are only run once, dependencies should be set only via `DependsOnTargets` or `<Target>DependsOn`.

### Actual Behavior

The target is run multiple times.

### Environment

dotnet 6.0.100
macos workload 12.0.101-preview.10.251

**Comment by chamons** [comment]
Setting as p3 as we're doing the right thing, but just slowly (if I understand the issue).

**Comment by rachelkang** [comment]
cc @rolfbjarne FYI

**Comment by rolfbjarne** [comment]
@Therzok are you building for a single runtime identifier, or multiple runtime identifiers?

**Comment by Therzok** [comment]
Single runtime identifier, `osx-x64`

**Comment by rolfbjarne** [comment]
> > @Therzok are you building for a single runtime identifier, or multiple runtime identifiers?
> 
> Single runtime identifier, `osx-x64`

No idea what's going on then, but a sample/binlog would be nice to have to track it down.

**Comment by Therzok** [comment]
Uploading a `no change` incremental build log:
https://microsofteur-my.sharepoint.com/:u:/g/personal/maungu_microsoft_com/EaMfdmRZ2JNNm3ymChStSKMBMrMWZn4rzdwx6eMna_wgfw?e=qyxuBJ

**Comment by rolfbjarne** [comment]
> Targets are only run once, dependencies should be set only via `DependsOnTargets` or `<Target>DependsOn`.

It seems that skipped targets aren't considered targets that have run, so they can be skipped multiple times.

Example:

```csproj
<?xml version="1.0" encoding="utf-8"?>
<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <ItemGroup>
        <Letters Include="nothing/**/*" />
    </ItemGroup>
    <Target Name="A" Inputs="@(Letters)" Outputs="A.something" />
    <Target Name="Build" DependsOnTargets="A;A" />
</Project>
```

```
$ msbuild test.proj /v:d /nologo         
Build started 12/16/2021 8:05:49 AM.
Process = "/Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64"
MSBuild executable path = "/Library/Frameworks/Mono.framework/Versions/6.12.0/lib/mono/msbuild/Current/bin/MSBuild.dll"
Command line arguments = "/Library/Frameworks/Mono.framework/Versions/6.12.0/lib/mono/msbuild/15.0/bin/MSBuild.dll test.proj /v:d /nologo"
Current directory = "/Users/rolf/test"
MSBuild version = "16.10.1"
Project "/Users/rolf/test/test.proj" on node 1 (default targets).
Building with tools version "Current".
Project file contains ToolsVersion="4.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="Current". For more information, please see http://go.microsoft.com/fwlink/?LinkId=293424.
Target "A" in project "/Users/rolf/test/test.proj" (target "Build" depends on it):
Skipping target "A" because it has no inputs.
Though the target has declared its inputs, the input specification only references empty properties and/or empty item lists.
Done building target "A" in project "test.proj".
Target "A" in project "/Users/rolf/test/test.proj" (target "Build" depends on it):
Skipping target "A" because it has no inputs.
Though the target has declared its inputs, the input specification only references empty properties and/or empty item lists.
Done building target "A" in project "test.proj".
Target "Build" in project "/Users/rolf/test/test.proj" (entry point):
Done building target "Build" in project "test.proj".
Done Building Project "/Users/rolf/test/test.proj" (default targets).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.25
```

So I don't think we can do anything here, it's just how MSBuild works.

**Comment by Therzok** [comment]
Technically, the targets could be made so the DependsOnTargets only appears once, as part of the DependsOn, like I mentioned in the issue description.

**Comment by Therzok** [comment]
Either DependsOnTargets is used on `_CoreCompileTextureAtlases` -> DependsOnTargets: `_ReadCoreCompileTextureAtlases` -> DependsOnTargets: `_BeforeCompileTextureAtlases;` or the `_CompileTextureAtlasesDependsOn` is the one driving the dependency chain, but I think the former version is the correct one.
